File: adventofcode2020/Day4.go
package main

import (
    "bufio"
    "fmt"
    "log"
	"os"
	"reflect"
	"sort"
)

func readFile(filename string)[]string{
	file, err := os.Open(filename)
    if err != nil {
        log.Fatal(err)
    }
    defer file.Close()

	var lines []string
    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
		line := scanner.Text()		
		lines = append(lines, line)
    }
    if err := scanner.Err(); err != nil {
        log.Fatal(err)
	}
	return lines
}

func findPoint(data string, ch int32, start int, end int) int{
	for _, c := range data {
		mid := (start+end)/2
		if c==ch{
			end = mid
		}else{
			start =mid+1
		}
		// fmt.Println(start, mid, end)
	}
	return start
}

func findAns(line string) int{
	row := findPoint(line[:7], 'F', 0, 127)
	col := findPoint(line[7:], 'L', 0, 7)
	// fmt.Println(row, col)
	return row*8+col
}

func main() {
	fmt.Println(reflect.TypeOf('F'))
	lines := readFile("input.txt")
	var max int
	var setids []int
	for _, line := range lines {
		ans := findAns(line)
		setids = append(setids, ans)
		if ans>max{
			max = ans
		}
	}
	fmt.Println(max)
	
	sort.Sort(sort.IntSlice(setids))
	// fmt.Println(setids)
	for i, setid := range setids{
		if i==0{
			continue
		}
		if setid-setids[i-1]>1{
			fmt.Println(i, setid)
		}
	}
}

File: advent-of-code/2022/day-4.py
inp = open("input.txt", "r")
lines = inp.readlines()
inputs = []

for line in lines:
    line = line.strip()
    inputs.append([int(val) for each in line.split(",") for val in each.split("-")])


def solution_one():
    ans = 0
    for line in inputs:
        if (line[0] <= line[2] and line[3] <= line[1]) or (
            line[2] <= line[0] and line[1] <= line[3]
        ):
            ans += 1

    return ans


def solution_two():
    ans = 0
    for line in inputs:
        if line[0] <= line[2] <= line[1] or line[0] <= line[3] <= line[1]:
            ans += 1
        elif line[2] <= line[0] <= line[3] or line[2] <= line[1] <= line[3]:
            ans += 1
    return ans


# print(f"Solution 1: {solution_one()}")
print(f"Solution 2: {solution_two()}")
